rfcs/0146-meta-categories.md

```diff
@@ -0,0 +1,284 @@
+
+1. A huge treewide edit of Nixpkgs
+
+   On the other hand, this is easily sprintable and amenable to automation.
+
+2. Bikeshedding
+
+   How many and which categories we should create? Can we expand them later?
+
+   For start, we can follow/take inspiration from many of the already existing
+   categories sets and add extra ones when the needs arise. Indeed, it is way
+   easier to create such categories using Nix language when compared to other
+   software collections.
+
+3. Superfluous
+
+   It can be argued that there are other ways to discover similar or related
+   package sets, like Repology.
+
+   However, this argument is a bit circular, because e.g. the classification
+   shown by Repology effectively replicates the classification done by the many
+   software collections in its catalog. Therefore, relying in Repology merely
+   transfers the question to external sources.
+
+   Further it becomes more pronounced when we take into account the fact Nixpkgs
+   is top 1 of most Repology statistics. The expected outcome, therefore, should
+   be precisely the opposite: Nixpkgs being _the_ source of structured metainfo
+   for other software collections.
+
+# Alternatives
+[alternatives]: #alternatives
+
+1. Do nothing
+
+   This will exacerbate the problems already listed.
+
+2. Ignore/nuke the categorization completely
+
+   This is an alternative worthy of some consideration. After all,
+   categorization is not without its problems, as shown above. Removing or
+   ignoring classification removes all problems.
+
+   However, there are good reasons to keep the categorization:
+
+   - The complete removal of categorization is too harsh. A solution that keeps
+     and enhances the categorization is way more preferrable than one that nukes
+     it completely.
+
+   - As said before, the categorization is already present; this RFC proposes to
+     expose it to a higher level, in a structured, more discoverable format.
+
+   - Categorization is very traditional among software collections. Many of them
+     are doing this just fine for years on end, and Nixpkgs can imitate them
+     easily - and even surpass them, given the benefits of Nix language
+     machinery.
+
+   - Categorization is useful in many scenarios and use cases - indeed they
+     are ubiquitous in software world:
+     - specialized search engines (from Repology to MELPA)
+     - code forges, from Sourceforge to Gitlab
+     - as said above, software collections from pkgsrc to slackbuilds
+     - to organization and preservation (as Software Heritage)
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Still unsolved is what data structure is better suited to represent a category.
+
+- For now we stick to a set `{ name, description }`.
+- Given the redundancy of the option above, another possibility is something
+  like `nameOfCategory = { description = ""; . . . }`
+
+# Future work
+[future]: #future-work
+
+- Curate the categories.
```

Comment:
This is imo the most important thing, a categorization is useless if it isn't curated. Unfortunately this is really not trivial, because both the set of packages and categories are really large, any package could be in any category, and both of them change over time.

So I would like to see some automated process for this, here's an idea for a very fancy one:
- When adding a package, one goes through a wizard-style questionnaire, answering questions about the package, ultimately determining whether each category applies, ideally as efficiently as possible.
- When making a proposal to add a new category, these steps need to be done:
  - Determine how it relates to other categories: Is it a sub-category, does it conflict with others?
  - Justify that the category is useful: Are there a bunch of packages it can be applied to?
  - Clearly defined the category, such that it's easy to determine for package maintainers whether the category applies
  - Somehow ask each package maintainer whether this new category applies to their package. This should be as simple to answer for maintainers as possible, e.g. just a single button.

I have a feeling that a Nixpkgs-specific website would be great to integrate this all together, along with E-Mails for notifications. Maybe this could be combined into e.g. a maximum once-per-week E-Mail, asking package maintainers to perform necessary actions to maintain their packages, which could also include things like reviewing relevant PR's or issues, tending to Hydra build failures, etc.

Reply:
1. This whole thing will require a team
2. Would this be the return of the mail-driven update requests? I really miss the Hydra "your package failed/passed" messages on my mailbox.

Reply:
I think we shouldn't reinvent the wheel and work as much as possible with existing categorization data. For example, many applications have desktop files whose categories (and tags) one could use. And maybe repology has something useful too

Reply:
I guess that _is_ a benefit of having to decide on _exactly one_ category for a package: Either you'll find the most accurate category, or none of them apply which justifies creating a new one. Therefore there is no general category maintenance required, it happens naturally as packages get added.

We should consider this as an alternative, something like
```nix
meta = {
  category = lib.categories.compiler;
}
```

And if we want to be more specific we could have category and subcategory, or something like that

Reply:
I feel this is not good, because it would defeat the cases of multi-valued categorization.

What about mail server, game server, file server...? One category for each looks too much.

Reply:
Yeah I agree it's not perfect, but if I had to choose between a curated single-valued categorization and an uncurated multi-valued one, I think I'd pick the former.

Reply:
At least the multi-valued version can include all that others (including upstreams, if they provide _some of the many_ categorisation labels) think too. Agreeing on the layout of anything single-valued will go as well as any attempt to globally agree on anything in Nixpkgs, it will be a weird and inconvenient to all (but in different ways, of course) compromise.

Reply:
> Yeah I agree it's not perfect, but if I had to choose between a curated single-valued categorization and an uncurated multi-valued one, I think I'd pick the former.

An uncurated multi-valued can possibly be curated in a mostly independent fashion - curating does not block a package's inclusion.

On the other hand, seeking the most fitting one-valued category puts a non-negligible mental load.